core/src/main/java/org/apache/calcite/rex/RexExecutable.java

```diff
@@ -17,100 +17,38 @@
-
-  private final Function1<DataContext, @Nullable Object @Nullable []> compiledFunction;
-  private final String code;
-  private @Nullable DataContext dataContext;
-
-  public RexExecutable(String code, Object reason) {
-    this.code = code;
-    this.compiledFunction = compile(code, reason);
-  }
-
-  private static Function1<DataContext, @Nullable Object @Nullable []> compile(String code,
-      Object reason) {
-    try {
-      final ClassBodyEvaluator cbe = new ClassBodyEvaluator();
-      cbe.setClassName(GENERATED_CLASS_NAME);
-      cbe.setExtendedClass(Utilities.class);
-      cbe.setImplementedInterfaces(new Class[] {Function1.class, Serializable.class});
-      cbe.setParentClassLoader(RexExecutable.class.getClassLoader());
-      cbe.cook(new Scanner(null, new StringReader(code)));
-      Class c = cbe.getClazz();
-      //noinspection unchecked
-      final Constructor<Function1<DataContext, @Nullable Object @Nullable []>> constructor =
-          c.getConstructor();
-      return constructor.newInstance();
-    } catch (CompileException | IOException | InstantiationException
-        | IllegalAccessException | InvocationTargetException
-        | NoSuchMethodException e) {
-      throw new RuntimeException("While compiling " + reason, e);
-    }
-  }
-
-  public void setDataContext(DataContext dataContext) {
-    this.dataContext = dataContext;
-  }
-
-  public void reduce(RexBuilder rexBuilder, List<RexNode> constExps,
-      List<RexNode> reducedValues) {
-    @Nullable Object[] values;
-    try {
-      values = execute();
-      if (values == null) {
-        reducedValues.addAll(constExps);
-        values = new Object[constExps.size()];
-      } else {
-        assert values.length == constExps.size();
-        final List<@Nullable Object> valueList = Arrays.asList(values);
-        for (Pair<RexNode, @Nullable Object> value : Pair.zip(constExps, valueList)) {
-          reducedValues.add(
-              rexBuilder.makeLiteral(value.right, value.left.getType(), true));
-        }
-      }
-    } catch (RuntimeException e) {
-      // One or more of the expressions failed.
-      // Don't reduce any of the expressions.
-      reducedValues.addAll(constExps);
-      values = new Object[constExps.size()];
-    }
-    Hook.EXPRESSION_REDUCER.run(Pair.of(code, values));
-  }
-
-  public Function1<DataContext, @Nullable Object @Nullable []> getFunction() {
-    return compiledFunction;
-  }
-
-  public @Nullable Object @Nullable [] execute() {
-    return compiledFunction.apply(requireNonNull(dataContext, "dataContext"));
-  }
+public interface RexExecutable {
+
+  /** Reduces expressions, and writes their results into {@code reducedValues}.
+   *
+   * <p>If an expression cannot be reduced, writes the original expression.
+   * For example, {@code CAST('abc' AS INTEGER)} gives an error when executed, so the executor
+   * ignores the error and writes the original expression.
+   *
+   * @param rexBuilder Rex builder
```

Comment:
Some functions are deleted, which may cause compatibility issues in the downstream project.  But I think it is accepted if the change is positive.